packages/transport/src/create.ts

```diff
@@ -1,122 +1,158 @@
-// NOTE: Code taken and modified from original: https://github.com/turbocrime/transport-demo
+import {
+  createRouterTransport,
+  ServiceImpl,
+  MethodImpl,
+  ConnectRouter,
+  ConnectError,
+  Code as ConnectErrorCode,
+} from '@connectrpc/connect';
+
+import {
+  MethodKind,
+  MethodInfo,
+  ServiceType,
+  AnyMessage,
+  JsonValue,
+  Any,
+} from '@bufbuild/protobuf';
 
-import { Code, ConnectError, createRouterTransport, ServiceImpl } from '@connectrpc/connect';
-import { MethodKind, ServiceType } from '@bufbuild/protobuf';
 import {
-  CreateAnyImplMethod,
-  GrpcRequest,
-  GrpcResponse,
-  isErrorResponse,
-  isServiceGrpcResponse,
-  isStreamResponse,
-  isUnaryResponse,
-  PendingRequests,
+  TransportMessage,
+  isTransportData,
+  InitChannelClientDataType,
+  InitChannelClientMessage,
 } from './types';
-import { unaryIO } from './unary';
-import { serverStreamIO } from './stream';
+
+import { JsonToMessage, streamToGenerator } from './stream';
+
 import { typeRegistry } from '@penumbra-zone/types/src/registry';
 
-// Creates a new service implementation by wrapping the original with the special event transport methods
-const makeAnyServiceImpl = <S extends ServiceType>(
-  service: ServiceType,
-  createMethod: CreateAnyImplMethod,
-): ServiceImpl<S> => {
-  const impl: ServiceImpl<typeof service> = {};
-  for (const [localName, methodInfo] of Object.entries(service.methods)) {
-    const method = createMethod({
-      ...methodInfo,
-      localName,
-      service,
-    });
-    if (method) impl[localName] = method;
+type CreateAnyMethodImpl<S extends ServiceType> = (
+  methodInfo: MethodInfo & { kind: MethodKind },
+) => typeof methodInfo extends S['methods'][keyof S['methods']]
+  ? MethodImpl<typeof methodInfo>
+  : never;
+
+type CreateAnyServiceImpl = <S extends ServiceType>(
+  service: S,
+  createMethod: CreateAnyMethodImpl<S>,
+) => Partial<ServiceImpl<S>>;
+
+const makeAnyServiceImpl: CreateAnyServiceImpl = <S extends ServiceType>(
+  service: S,
+  createMethod: CreateAnyMethodImpl<S>,
+): Partial<ServiceImpl<S>> => {
+  const impl = {} as { [P in keyof S['methods']]: MethodImpl<S['methods'][P]> };
+  let localName: keyof S['methods'];
+  let methodInfo: MethodInfo;
+  for ([localName, methodInfo] of Object.entries(service.methods))
+    impl[localName] = createMethod(methodInfo);
+  return impl;
+};
+
+type Promised<T> = Parameters<ConstructorParameters<typeof Promise<T>>[0]>;
```

Comment:
Is this grabbing the parameter of the resolve function? I wonder if we can name this something more clear on what it is.